### PR TITLE
Add Parity (Ethereum client)

### PR DESCRIPTION
**Pull request by dbrock**
###### Motivation for this change

Add `parity`, the new, fast, lightweight Ethereum client. (We already have `go-ethereum`.) https://ethcore.io/parity
###### Things done
- [ ] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [x] NixOS
  - [ ] OS X
  - [ ] Linux
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dbrock** [comment]
I guess that `parity` build may not be deterministic...

**Comment by dbrock** [comment]
Hmm, it seems to build consistently on my machine though... I wonder what's up with Travis.

Waiting for @mbrock to try it out.

**Edit:** Just realized I had to turn off sandboxing for it to build, because otherwise I got messages to the effect of "couldn't find host github.com" --- some kind of DNS configuration problem, I suppose...?

**Comment by globin** [comment]
Builds are not allowed to access the network in the sandbox.

**Comment by the-kenny** [comment]
It's generally a good idea to keep the sandbox enabled during builds. If your package doesn't build with sandboxing enabled, it usually has some kind of impurity.

**Comment by dbrock** [comment]
Ah, okay, that explains a lot, thanks guys.

Well, ok, we might have to wait for the next official release of Parity then before this can be merged. (I'm guessing that's why it's going to the network... this development version probably has dependencies on like the HEADs of other repositories, or something like that.)

The current version (1.2.4) unfortunately seems to depend on a broken (https://github.com/facebook/rocksdb/issues/1117) RocksDB.

cc @gavofyork if interested

**Comment by dbrock** [comment]
So they actually just released Parity 1.3.0 last night, and the RocksDB issue is now fixed.

Unfortunately, it still does not build in the sandbox, as it still attempts to access github.com (see https://github.com/ethcore/parity/search?utf8=%E2%9C%93&q=github).

```
unpacking source archive /nix/store/r55an1pxd43dlnw5j08qr50618xhbd1y-v1.3.0.tar.gz
source root is parity-1.3.0
Using cargo deps from /nix/store/vkr2f5dls3lbc949fy4ph4qniipf1a2w-parity-1.3.0-fetch
Using indexHash '-ba82b75dd6681d6f'
Using rust registry from /nix/store/2n8czavmwlb7dmvllv6z0dax1lgrl0s3-rustRegistry-2016-08-10-fe018be
warning: custom registry support via the `registry.index` configuration is being removed, this functionality will not work in the future
    Updating git repository `https://github.com/ethcore/rust-ctrlc.git`
warning: spurious network error (2 tries remaining): [12/-1] curl error: Couldn't resolve host 'github.com'

warning: spurious network error (1 tries remaining): [12/-1] curl error: Couldn't resolve host 'github.com'

error: failed to load source for a dependency on `ctrlc`

Caused by:
  Unable to update https://github.com/ethcore/rust-ctrlc.git#f4927770

Caused by:
  failed to fetch into /tmp/nix-build-parity-1.3.0.drv-0/deps/git/db/rust-ctrlc-9a5ce78f777d3f83
```

Does anyone know how to fix this? I tried using
```
  src = fetchurl {
    url = "https://github.com/ethcore/parity/archive/v${version}.tar.gz";
    sha256 = "07n2s29iyjp112shq1mphy1mglb829q3jblk06i6abba234c7dm2";
  };
```
instead of `fetchFromGitHub`, but got the same result.

**Comment by the-kenny** [comment]
That happens because the `Cargo.toml` file of parity specified a custom URL for a dependency at https://github.com/ethcore/parity/blob/master/Cargo.toml#L29

Unfortunately I don't know how to work around this - you need to fetch this dependency manually via `fetchgit` or similar and put it somehow in the right place.

**Comment by ryepdx** [comment]
Looks like cargo supports [implicit local package overrides via magic paths](http://doc.crates.io/specifying-dependencies.html#overriding-with-local-dependencies). Could you leverage that to override the Github dependencies? Not familiar with Nix script yet, so I don't actually know what's possible.

**Comment by the-kenny** [comment]
The feature you linked might work for this - I'll give it a try tomorrow (too tired right now, sorry).

Note to self: Implement this in a generic fashion, it will also help other packages (like rustup.rs)

File: pkgs/applications/altcoins/mist-browser.nix [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  _name = "Mist-${platform}-${_version}";
+  _version = builtins.replaceStrings ["."] ["-"] version;
+
+  src = fetchurl {
+    url = "https://github.com/ethereum/mist/releases/download/v${version}/${_name}.deb";
+    sha256 = "0kyszkjqhas74gazr80y853z2nycn6fh2mysn5pplpqkhmpz3yl8";
+  };
+
+  icon = fetchurl {
+    url = "https://raw.githubusercontent.com/ethereum/mist/master/icons/wallet/icon.png";
```
Comment (the-kenny) [inline pkgs/applications/altcoins/mist-browser.nix]: This will break as soon someone updates the icon in master.

File: pkgs/applications/altcoins/mist-browser.nix [hunk 2]
```diff
@@ -0,0 +1,81 @@
+{ pkgs, stdenv, alsaLib, atk, cairo, cups, dbus, dpkg, expat, fetchurl,
+fontconfig, freetype, gdk_pixbuf, gnome, glib, gtk, go-ethereum, libcap,
```
Comment (the-kenny) [inline pkgs/applications/altcoins/mist-browser.nix]: Please take a look at the coding style guidelines at https://nixos.org/nixpkgs/manual/#chap-conventions

File: pkgs/applications/altcoins/mist-browser.nix [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        xorg.libXtst xorg.libXcomposite xorg.libXi xorg.libXfixes
+        xorg.libXrandr xorg.libXcursor xorg.libXScrnSaver
+      ];
+in stdenv.mkDerivation rec {
+  name = "mist-browser-${version}";
+  version = "0.8.3";
+
+  is64Bit = (stdenv.system == "x86_64-linux");
+  platform = (if is64Bit then "linux64" else "linux32");
+  _name = "Mist-${platform}-${_version}";
```
Comment (the-kenny) [inline pkgs/applications/altcoins/mist-browser.nix]: Can you use a more descriptive name? `upstream_name` or something

File: pkgs/applications/altcoins/parity.nix [hunk 4]
```diff
@@ -0,0 +1,22 @@
+{ stdenv, fetchurl, rustPlatform }:
+
+with rustPlatform;
+
+buildRustPackage rec {
+  name = "parity-${version}";
+  version = "1.3.0";
+
+  src = fetchurl {
```
Comment (the-kenny) [inline pkgs/applications/altcoins/parity.nix]: Better use `fetchFromGitHub`

File: pkgs/top-level/all-packages.nix [hunk 5]
```diff
@@ -10497,7 +10497,7 @@ in
 
   influxdb = (callPackage ../servers/nosql/influxdb/v0.nix { }).bin // { outputs = [ "bin" ]; };
 
-  influxdb10 = (callPackage ../servers/nosql/influxdb/v1.nix { }).bin // { outputs = [ "bin" ]; }; 
+  influxdb10 = (callPackage ../servers/nosql/influxdb/v1.nix { }).bin // { outputs = [ "bin" ]; };
```
Comment (the-kenny) [inline pkgs/top-level/all-packages.nix]: Why are these whitespace changes in your PR?

File: pkgs/tools/misc/unclutter-xfixes/default.nix [hunk 6]
```diff
@@ -7,7 +7,7 @@ let version = "1.2"; in
 stdenv.mkDerivation {
   name = "unclutter-xfixes-${version}";
   version = version;
-
```
Comment (the-kenny) [inline pkgs/tools/misc/unclutter-xfixes/default.nix]: Another unrelated whitespace-change. In a project the size of nix we need to be careful with those as they easily cause merge conflicts. 

(I'm all :+1: on removing trailing whitespace though!)

File: nixos/modules/services/mail/opensmtpd.nix [hunk 7]
```diff
@@ -109,13 +109,15 @@ in {
       after = [ "network.target" ];
       preStart = ''
         mkdir -p /var/spool/smtpd
+        chmod 711 /var/spool/smtpd
```
Comment (the-kenny) [inline nixos/modules/services/mail/opensmtpd.nix]: These changes don't seem to belong to this PR

**Comment by dbrock** [comment]
I'm sorry, something weird seems to have happened here. I think @ryepdx is pushing changes to this branch not realizing it is linked to this pull request.

It's my fault for accidentally starting the PR from the master branch and then not being able to change it. (There's no way to change the branch of a PR, right?)

@ryepdx: please avoid using the `master` branch on `nexusdev/nixos` until this PR has been merged.

Or we could just close this PR and I'll start a new one from a better branch name?

**Comment by the-kenny** [comment]
Feel free to open a new one. As long as the final PR is clean I don't care for the branch name :)
